src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -0,0 +1,422 @@
+                }
+
+                _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+
+                // Cancel any remaining asynchronous work we may have had to update this file
+                _cancellationTokenSource.Cancel();
+            }
+
+            public async Task UpdateBufferContentsAsync(CancellationToken cancellationToken)
+            {
+                await TaskScheduler.Default;
+
+                GeneratedSourceResult? generatedSource = null;
+                var project = _workspace.CurrentSolution.GetProject(_projectId);
+
+                // Locals correspond to the equivalently-named fields; we'll assign these and then assign to the fields while on the
+                // UI thread to avoid any potential race where we update the InfoBar while this is running.
+                string? windowFrameMessageToShow;
+                ImageMoniker windowFrameImageMonikerToShow;
+
+                if (project == null)
+                {
+                    windowFrameMessageToShow = "The project no longer exists.";
+                    windowFrameImageMonikerToShow = KnownMonikers.StatusError;
+                }
+                else
+                {
+
+                    var generatorDriverRunResults = await project.GetGeneratorDriverRunResultAsync(cancellationToken).ConfigureAwait(false);
+                    var generatorRunResult = generatorDriverRunResults?.Results.SingleOrNull(r => r.Generator.GetType().FullName == _generatorTypeName);
+
+                    if (generatorRunResult == null)
+                    {
+                        windowFrameMessageToShow = string.Format(ServicesVSResources.The_generator_0_that_generated_this_file_has_been_removed_from_the_project, _generatorTypeName);
+                        windowFrameImageMonikerToShow = KnownMonikers.StatusError;
+                    }
+                    else
+                    {
+                        generatedSource = generatorRunResult.Value.GeneratedSources.SingleOrNull(r => r.HintName == _generatedSourceHintName);
+
+                        if (generatedSource == null)
+                        {
+                            windowFrameMessageToShow = string.Format(ServicesVSResources.The_generator_0_that_generated_this_file_has_stopped_generating_this_file, _generatorTypeName);
+                            windowFrameImageMonikerToShow = KnownMonikers.StatusError;
+                        }
+                        else
+                        {
+                            windowFrameMessageToShow = string.Format(ServicesVSResources.This_file_is_autogenerated_by_0_and_cannot_be_edited, _generatorTypeName);
+                            windowFrameImageMonikerToShow = default;
+                        }
+                    }
+                }
+
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
This pairing is unnecessary.

```suggestion
                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```